src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -109,42 +112,61 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
        shapes_it != tasks_to_dispatch_.end();) {
     auto &dispatch_queue = shapes_it->second;
     for (auto work_it = dispatch_queue.begin(); work_it != dispatch_queue.end();) {
-      auto work = *work_it;
-      auto task = std::get<0>(work);
-      auto spec = task.GetTaskSpecification();
+      auto &work = *work_it;
+      auto &task = std::get<0>(work);
+      auto &spec = task.GetTaskSpecification();
 
       std::shared_ptr<WorkerInterface> worker = worker_pool.PopWorker(spec);
       if (!worker) {
         // No worker available, we won't be able to schedule any kind of task.
         return;
       }
 
-      std::shared_ptr<TaskResourceInstances> allocated_instances(
-          new TaskResourceInstances());
-      bool schedulable = cluster_resource_scheduler_->AllocateLocalTaskResources(
-          spec.GetRequiredResources().GetResourceMap(), allocated_instances);
-      if (!schedulable) {
-        // Not enough resources to schedule this task.
+      const auto owner_worker_id = WorkerID::FromBinary(spec.CallerAddress().worker_id());
+      const auto owner_node_id = NodeID::FromBinary(spec.CallerAddress().raylet_id());
+      // If the owner has died since this task was queued, cancel the task by
+      // killing the worker (unless this task is for a detached actor).
+      if (!worker->IsDetachedActor() &&
+          !is_owner_alive_(owner_worker_id, owner_node_id)) {
+        // TODO (Alex): To @swang pls double check and make sure this is ok. <3 thx
+        RAY_LOG(ERROR) << "Owner of assigned task "
+                       << task.GetTaskSpecification().TaskId()
+                       << " died, killing leased worker " << worker->WorkerId();
         worker_pool.PushWorker(worker);
-        // All the tasks in this queue are the same, so move on to the next queue.
-        break;
-      }
-
-      auto reply = std::get<1>(work);
-      auto callback = std::get<2>(work);
-      worker->SetOwnerAddress(spec.CallerAddress());
-      if (spec.IsActorCreationTask()) {
-        // The actor belongs to this worker now.
-        worker->SetLifetimeAllocatedInstances(allocated_instances);
       } else {
-        worker->SetAllocatedInstances(allocated_instances);
-      }
-      worker->AssignTaskId(spec.TaskId());
-      if (!RayConfig::instance().enable_multi_tenancy()) {
-        worker->AssignJobId(spec.JobId());
+        std::shared_ptr<TaskResourceInstances> allocated_instances(
```

Comment:
This function is getting a bit long. Do you think we could separate some of the logic in this if-else into separate functions?

Reply:
yeah i kinda agree, the diff will just grow uglier

Reply:
Ok this pretty much had to happen to handle your merge conflict.